integrations/integration.php

```diff
@@ -1,55 +1,260 @@
 <?php
+/**
+ * Base class for Integration.
+ *
+ * @package Automattic\VIP\Integrations
+ */
 
 namespace Automattic\VIP\Integrations;
 
+use InvalidArgumentException;
+
+// phpcs:disable Generic.Files.OneObjectStructurePerFile.MultipleFound -- Disabling due to enums.
+
+/**
+ * Enum which represent all possible statuses for the client integration via VIP.
+ *
+ * These should be in sync with the statuses available on the backend.
+ */
+abstract class Client_Integration_Status {
+	const BLOCKED = 'blocked';
+}
+
+/**
+ * Enum which represent all possible statuses for the site integration via VIP.
+ *
+ * These should be in sync with the statuses available on the backend.
+ */
+abstract class Site_Integration_Status {
+	const ENABLED  = 'enabled';
+	const DISABLED = 'disabled';
+	const BLOCKED  = 'blocked';
+}
+
 /**
  * Abstract base class for all integration implementations.
  *
  * @private
  */
 abstract class Integration {
+	/**
+	 * Slug of the integration.
+	 *
+	 * @var string
+	 */
+	private string $slug;
+
 	/**
 	 * An optional configuration array for this integration, added during activation.
 	 *
 	 * @var array
 	 */
-	protected array $config = [];
+	private array $customer_config = [];
```

Comment:
Is there a way we could combine this, and `vip_config` into one array? That way, each integration implementation has only a single config to consume and work on. In the case where the integration is implemented via `mu-plugins`, the config would be the same as what $customer_config would be. If it's via the other route, then it would realistically just be vip_config and not customer_config as well.

Reply:
This is what my understanding is:

- `customer_config` are the configs provided by customer for plugin configuration and it can include things like versioning and custom plugin settings.

- `vip_config` are the configs which are provided by VIP and it can include things like status/secrets of the integration (for client, site and network sites) and in future may include custom plugin settings. It have a relatively complex structure than `customer_config`

So I think for now it is a better decision to separate them because of different attributes and different structures of these configs. As of now from code we can easily make the decision to handle `customer_config` easily and handle `vip_config` with care because it contains sensitive info but if we mix these then I think it can introduce complexity which will be difficult to understand and maintain.

Reply:
In the original Integrations PR, we'd talked about potentially keeping things simple and combining both activation paths, which I think we should do.

See:
- https://github.com/Automattic/vip-go-mu-plugins/pull/4458#discussion_r1220373639
- https://github.com/Automattic/vip-go-mu-plugins/pull/4458#issuecomment-1544102329

Here's one way we could do this:

```
diff --git a/integrations/integrations.php b/integrations/integrations.php
index 35bc43a6a..5712486f5 100644
--- a/integrations/integrations.php
+++ b/integrations/integrations.php
@@ -54,6 +54,15 @@ class Integrations {
 		return $this->integrations[ $slug ] ?? null;
 	}
 
+	/**
+	 * Returns all registered integrations.
+	 *
+	 * @return array<Integration>
+	 */
+	public function get_registered(): array {
+		return $this->integrations;
+	}
+
 	/**
 	 * Call load() for each registered and activated integration.
 	 *
diff --git a/vip-integrations.php b/vip-integrations.php
index 0728a6326..6e9ff5127 100644
--- a/vip-integrations.php
+++ b/vip-integrations.php
@@ -34,6 +34,17 @@ function activate( string $slug, array $config = [] ): void {
 // Load integrations in muplugins_loaded:5 to allow integrations to hook
 // muplugins_loaded:10 or any later action.
 add_action( 'muplugins_loaded', function() {
+	foreach ( IntegrationsSingleton::instance()->get_registered() as $slug => $integration ) {
+		if ( $integration->is_active_via_config_file() ) {
+			$config = $integration->get_config_from_config_file();
+
+			activate( $slug, array(
+				'loaded_via' => 'platform-config',
+				'config' => $config,
+			) );
+		}
+	}
+
 	IntegrationsSingleton::instance()->load_active();
 }, 5 );
```

The nice thing is that we have a consistent pathway for activation. I think there's some limits here in that customers calling `activate` directly can't as easily set configs for network site, etc. but we could make that an explicit limitation.

Reply:
Currently keys of `$customer_config` param aren't defined yet but most probably it will include `$version` and maybe some more things and `$vip_config` support following params

**Note:** As of versioning of `wp-parsely` is supported via`wpvip_parsely_version` filter and we are not changing it.

```
{
  "client": {
    "status": ""
  },
  "site": {
    "status": "",
    "config": {}
  },
  "network_sites": {
    "1": {
      "status": "",
      "config": {}
    }
  }
}
```

So can you please suggest how should I extend this? I am thinking of extending the config provided by platform with the customer configs and then forward it it to the integration (i.e. plugin) via `setup_config` method .... in this case customer configs will take priority .... is it okay?

It will be helpful if we can decides what kind of values we could expect from `customer_configs` ..... I think if its generic across all integrations then it does make sense to use one config and one filter but if could going to be different for each integration then maybe its better to keep this separate (personally I prefers to keep it generic across all integrations but let me know about your thoughts)

Reply:
Addressed in https://github.com/Automattic/vip-go-mu-plugins/pull/4660/commits/0f6f3967c49b9cd9e84a6432566b2a1d81a14417 (tests are updated seprately)